### ISSUE TITLE
[Feature Request] Implement activity task poll limit

### ISSUE THREAD

**Issue by madhav2302** [temporalio/features#443]
### Is your feature request related to a problem? Please describe.

It would be nice number of times an activity worker can poll tasks, as our service is currently not multi-tenant, and we would like to set it to 1. It will also help us to shutoff the workers gracefully by limiting the number of activities we allow it to run, as our activities can be long running, and we can't kill the activity and start new as progress will be lost. 

### Describe the solution you'd like

It can be handled with 2 solutions: 
- Poller suspend polling after polling a task successfully
- Provide MaximumPollTaskPerPoller. See implementation at https://gist.github.com/madhav2302/e35a9336b2fd8ccfc36fe4452d9c26b5

**Comment by madhav2302** [comment]
@cretz I tried it. 

- worker shutdown shuts down the heartbeat executor as well, so it doesn't work for us.
- stopping pollers have some delay, as it waits for activity to start. For quick activities poller makes the call to temporal server to fetch next task.

**Comment by amanpsbhatia** [comment]
+1 This would be great feature to have.

**Comment by cretz** [comment]
> worker shutdown shuts down the heartbeat executor as well, so it doesn't work for us.

I think this is a Java SDK thing. The way worker shutdown works in most SDKs is to stop polling and, with optional grace period, cancel activities and wait indefinitely for them to complete (while still letting heartbeat work). But Java SDK has the stop-polling step separate so can use just that.

> stopping pollers have some delay, as it waits for activity to start. For quick activities poller makes the call to temporal server to fetch next task.

But it won't fetch the next task if you have max activities as 1 and you are still _in_ that 1 activity, correct? It should only poll again if you have available slots, which you don't until you have stopped polling.

**Comment by madhav2302** [comment]
> It should only poll again if you have available slots, which you don't until the activity completes, but you have since stopped polling before that next poll could start. (I have not tested this)

I don't see the logic in the java SDK where it checks for available slots. Maybe that is missing, that's why polling doesn't stop. 

Can you confirm if Java SDK checks available slots before polling? if not, it would be nice to implement that.

**Comment by cretz** [comment]
:+1: Will confer with Java SDK owner next week on this to confirm (cc @Quinn-With-Two-Ns). Basically if you put `setMaxConcurrentActivityExecutionSize` as `1` on `WorkerOptions.Builder`, then that worker will not even ask for a second item of work while the first is running.

**Comment by madhav2302** [comment]
@cretz yes, we have done this.  See below -

```
            var workerOptions =
                    WorkerOptions.newBuilder(WorkerOptions.getDefaultInstance())
                            .setMaxConcurrentActivityExecutionSize(1)
                            .setMaxConcurrentActivityTaskPollers(1)
                            .build();
```

It still isn't stopping the polling when we don't have available slots.

**Comment by madhav2302** [comment]
@cretz @Quinn-With-Two-Ns I wrote a test at https://gist.github.com/madhav2302/67de7f577f9be1f2592193982b5a4fa5, where 2nd activity runs on a worker even after polling is suspended. 

@cretz as you said there is a check for available slots [here](https://github.com/temporalio/sdk-java/blob/0c6c566a89f28f9a7715fcb258c3c2401ea310e7/temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityPollTask.java#L97-L102) as a semaphore lock, but after this we don't check if polling is suspended, which results in activity execution on worker.

**Comment by Quinn-With-Two-Ns** [comment]
@cretz is correct `setMaxConcurrentActivityExecutionSize` as 1 will limit the worker to executing 1 normal activity at a time.

@madhav2302 That would be expected `suspendPolling` suspends any new worker poll requests, but currently outstanding ones could still return workflow/activity tasks after its called and those tasks will be run.

**Comment by madhav2302** [comment]
@Quinn-With-Two-Ns is it expected behavior to do one last poll, even after we have suspended polling? it doesn't look correct, as after polling is suspended, no single activity run should happen.

**Comment by Quinn-With-Two-Ns** [comment]
>as after polling is suspended, no single activity run should happen.

This is not true, after suspended polling is called any outstanding poll request can still be severed. The same thing is also explained in this forum post https://community.temporal.io/t/worker-suspendpolling-not-working/7102

**Comment by madhav2302** [comment]
@Quinn-With-Two-Ns how would you suggest that 2nd activity mentioned in https://gist.github.com/madhav2302/67de7f577f9be1f2592193982b5a4fa5 doesn't run at all? 

or does it make sense to have this feature, or maybe add a check [here](https://github.com/temporalio/sdk-java/blob/0c6c566a89f28f9a7715fcb258c3c2401ea310e7/temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityPollTask.java#L97-L102) which checks if polling is suspended.

**Comment by Quinn-With-Two-Ns** [comment]
So I think the way to do it right now is shutdown the worker as the last thing you do in your activity, probably in a finally block. That would stop polling for new work before the slot is released by the activity finishing.

We can also check if we paused polling after getting a slot as well. This whole part of the SDK is being actively refactored and maybe we can changes this as part of that work. @Sushisource I think when we add the pauseable slot supplier wrapper we should make sure we check if it is paused before and after getting a slot in case getting a slot blocks and the pauseable state changes

**Comment by madhav2302** [comment]
> So I think the way to do it right now is shutdown the worker as the last thing you do in your activity, probably in a finally block. That would stop polling for new work before the slot is released by the activity finishing.

Ok, I will try that. 



> We can also check if we paused polling after getting a slot as well. This whole part of the SDK is being actively refactored and maybe we can changes this as part of that work.

That would be great. Can we add the issue in the refactor SDK parent, so it is not missed. Any idea when the refactored SDK will be released?

**Comment by madhav2302** [comment]
> We can also check if we paused polling after getting a slot as well. This whole part of the SDK is being actively refactored and maybe we can changes this as part of that work.

@Quinn-With-Two-Ns @Sushisource we can contribute it via a PR if it helps to get the fix released quickly.

**Comment by Sushisource** [comment]
> We can also check if we paused polling after getting a slot as well. This whole part of the SDK is being actively refactored and maybe we can changes this as part of that work. @Sushisource I think when we add the pauseable slot supplier wrapper we should make sure we check if it is paused before and after getting a slot in case getting a slot blocks and the pauseable state changes

Pausing the supplier is, in effect, making the slot reservation call block until un-paused. If you pause the handing out of slots _just after_ a slot was handed out, well, you handed it out, and it's going to get used.

The desired semantic is something more like "invalidate any issued but un-used slots and also cancel any in-flight poll RPCs". And that, ultimately, is pretty much shutdown - just without the minor overhead of restarting the worker later.

So I'm actually not sure what the right default semantic for a `PauseableSlotSupplier` would be. I can buy that the right default behavior might actually be the semantic I just described, rather than the more literal "pause handing out of any new slots". Maybe we want both, but I think that's starting to get a bit confusing.

@madhav2302 The good news for you is we'll be releasing this interface fairly soon, and you can implement it to do 90% of what you want - but this particular race may not be addressable in the first cut. We'll discuss internally if we want to add this particular ability in the first preview. No need for contribution, but thank you!

**Comment by madhav2302** [comment]
> The good news for you is we'll be releasing this interface fairly soon, and you can implement it to do 90% of what you want - but this particular race may not be addressable in the first cut. We'll discuss internally if we want to add this particular ability in the first preview. No need for contribution, but thank you!

@Sushisource any updates on it?

**Comment by pdaga2** [comment]
+1
